src/libmongoc/src/mongoc/mongoc-crypt.c

```diff
@@ -575,6 +576,35 @@ _state_need_kms (_state_machine_t *state_machine, bson_error_t *error)
 #undef BUFFER_SIZE
 }
 
+static bool
+_state_need_kms_credentials (_state_machine_t *sm, bson_error_t *error)
+{
+   bson_t creds = BSON_INITIALIZER;
+   BSON_ASSERT (sm->crypt->creds_cb.fn);
```

Comment:
Should this assertion be converted into a "nice" BSON error instead? Given the possibility (is there?) that the `NEED_KMS_CREDENTIALS` state is reached without a callback being set (the default). If we want to enforce this assertion, I suggest adding a not-null assertion to `mongoc_auto_encryption_opts_set_kms_credential_provider_callback`.

Reply:
In its current state, the code only gets here if the callback is non-null. This state needs to be explicitly enabled when the context was created.